File: ChinoKafuu/Engine/Applio/TTSApi.py
from gradio_client import Client,handle_file
import os

client = Client("http://127.0.0.1:6969/")

def TTS(message, output_folder="resultfolder" ,output_file="result.wav"):
    os.makedirs(output_folder, exist_ok=True)
    full_output_path = os.path.join(output_folder, output_file)
    
    result = client.predict(
		tts_text=message,
		tts_voice="ja-JP-NanamiNeural",
		tts_rate=0,
		pitch=3,
		filter_radius=4,
		index_rate=0.6,
		volume_envelope=1,
		protect=0.5,
		hop_length=256,
		f0_method="rmvpe",
		output_tts_path="result",
		output_rvc_path=full_output_path,
		pth_path="logs\chino-kafuu\chino-kafuu.pth",
		index_path="logs\chino-kafuu\added_IVF209_Flat_nprobe_1_chino-kafuu_v2.index",
		split_audio=False,
		f0_autotune=True,
		clean_audio=True,
		clean_strength=0.5,
		export_format="WAV",
		upscale_audio=False,
		f0_file=handle_file('https://github.com/gradio-app/gradio/raw/main/test/test_files/sample_file.pdf'),
		embedder_model="contentvec",
		embedder_model_custom=None,
		api_name="/run_tts_script"
	)
    return result

File: ChinoKafuu/Engine/Applio/TTSApi.cs
using System.Net.Http.Json;
using System.Text.Json.Serialization;

public class TTSRequest
{
    [JsonPropertyName("guild_id")]
    public string guild_id { get; set; }
    
    [JsonPropertyName("text")]
    public string Text { get; set; }
    
    [JsonPropertyName("voice")]
    public string Voice { get; set; } = "ja-JP-NanamiNeural";
    
    [JsonPropertyName("rate")]
    public int Rate { get; set; } = 0;
    
    [JsonPropertyName("pitch")]
    public int Pitch { get; set; } = 4;
    
    [JsonPropertyName("index_rate")]
    public float IndexRate { get; set; } = 0.5f;
    
    [JsonPropertyName("volume_envelope")]
    public float VolumeEnvelope { get; set; } = 1.0f;
    
    [JsonPropertyName("protect")]
    public float Protect { get; set; } = 0.4f;
    
    [JsonPropertyName("f0_method")]
    public string F0Method { get; set; } = "rmvpe";
    
    [JsonPropertyName("split_audio")]
    public bool SplitAudio { get; set; } = true;
    
    [JsonPropertyName("f0_autotune")]
    public bool F0Autotune { get; set; } = false;
    
    [JsonPropertyName("f0_autotune_strength")]
    public float F0AutotuneStrength { get; set; } = 0.0f;
    
    [JsonPropertyName("clean_audio")]
    public bool CleanAudio { get; set; } = true;
    
    [JsonPropertyName("clean_strength")]
    public float CleanStrength { get; set; } = 0.2f;
    
    [JsonPropertyName("export_format")]
    public string ExportFormat { get; set; } = "wav";
    
    [JsonPropertyName("embedder_model")]
    public string EmbedderModel { get; set; } = "contentvec";
    
    [JsonPropertyName("embedder_model_custom")]
    public string? EmbedderModelCustom { get; set; } = null;
    
    [JsonPropertyName("gpu")]
    public int Gpu { get; set; } = 0;
    
    [JsonPropertyName("cache_data_in_gpu")]
    public bool CacheDataInGpu { get; set; } = true;
}

public class TTSApi
{
    private readonly HttpClient _httpClient;
    private const string BASE_API_URL = "http://localhost:8000";

    public TTSApi(HttpClient httpClient)
    {
        _httpClient = httpClient;
        _httpClient.Timeout = TimeSpan.FromMinutes(5);
    }

    public async Task<string> GenerateTTS(string message, string guildId, CancellationToken cancellationToken = default)
    {
        var ttsRequest = new TTSRequest
        {
            Text = message,
            guild_id = guildId
        };
        HttpResponseMessage response = await _httpClient.PostAsJsonAsync($"{BASE_API_URL}/tts", ttsRequest, cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            string responseContent = await response.Content.ReadAsStringAsync(cancellationToken);
            throw new Exception($"Gọi api thất bại: {response.ReasonPhrase} - {responseContent}");
        }

        var result = await response.Content.ReadFromJsonAsync<Dictionary<string, string>>(cancellationToken: cancellationToken);
        return result["file_name"];
    }

    public async Task DownloadGeneratedTTS(ulong guildId, string fileName, string outputPath, CancellationToken cancellationToken = default)
    {
        string outputDirectory = Path.GetDirectoryName(outputPath);

        Directory.CreateDirectory(outputDirectory);

        string decodedFileName = System.Web.HttpUtility.UrlDecode(fileName);
        string downloadUrl = $"{BASE_API_URL}/get-generated/{guildId}/{decodedFileName}";

        HttpResponseMessage response = await _httpClient.GetAsync(downloadUrl, cancellationToken);

        if (!response.IsSuccessStatusCode)
        {
            throw new Exception($"Tải file thất bại: {response.ReasonPhrase}");
        }

        await using FileStream fs = new FileStream(outputPath, FileMode.Create, FileAccess.Write, FileShare.None);
        await response.Content.CopyToAsync(fs, cancellationToken);
    }
}
